File: test/execution_example/models/smart_forecast_julia/smart_forecast.jl
using DataFrames
using Statistics
using Dates
using CSV
using StatsModels
using LinearRegression


time_series_path = ARGS[1]
num_steps_ahead = parse(Int, ARGS[2])
num_scenarios = parse(Int, ARGS[3])

# charger le dataframe
df = CSV.read(time_series_path, DataFrame)

HORIZON = num_steps_ahead  # Predict the length of the test dataset

# extraire l'année de chaque date
df[!, "annee"] = year.(df[!, "dates"])

# trouver la valeur maximale de chaque année et calculer la proportion pour chaque mois
max_values = combine(groupby(df, :annee), names(df[:, 2:end]) .=> maximum .=> names(df[:, 2:end]))

proportions = df[:, 2:5] ./ repeat(max_values[:, 2:5], inner=12)[1:size(df, 1), :]

# calculer la moyenne des proportions pour chaque mois
proportions[!, "dates"] = df[!, "dates"]
proportions[!, "month"] = month.(proportions[!, "dates"])
means = combine(groupby(proportions, :month), 
                :NORDESTE => mean,
                :SUDESTE => mean,
                :NORTE => mean,
                :SUL => mean, 
                renamecols=false)

# préparer les données pour la régression linéaire
X = convert(Array, max_values[:, 1])

degree = 2
X_poly = hcat([X.^d for d in 0:degree]...)

predicted_max_values = Vector{Any}()

for year in 0:trunc(Int, ceil(num_steps_ahead / 12)) - 1
    year_value = Vector{Any}()
    for region_nb in 2:5
        y_region = convert(Array, max_values[:, region_nb])
        lr = linregress(X_poly, y_region)
        max_values_pred = maximum(X) + 1 + year
        X_pred = hcat([max_values_pred.^d for d in 0:degree]...)
        max_value = lr(X_pred)
        push!(year_value, max_value[1])
    end
    push!(predicted_max_values, year_value)
end

predicted_max_values = rename(first(repeat(DataFrame(hcat(predicted_max_values...)', :auto), inner = 12), num_steps_ahead),  [:NORDESTE, :SUDESTE, :NORTE, :SUL])

means_values = means[!, 2:5]
means_values = repeat(means_values, outer=Int(ceil(num_steps_ahead / 12))) |> DataFrame

forecast_df = predicted_max_values .* means_values

dates = [df.dates[end-HORIZON] + Dates.Month(HORIZON) + Month(i-1) for i in 1:HORIZON]
scenario = fill(1, nrow(forecast_df))

insertcols!(forecast_df, 1, :dates=>dates)
insertcols!(forecast_df, 2, :scenario=>scenario)

output_path = joinpath(dirname(time_series_path), "_simulated_scenarios.csv")

# Write predictions to a CSV file
CSV.write(output_path, forecast_df, writeheader=true)

File: test/execution_example/models/smart_forecast_python/smart_forecast.py
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
import os
import sys
import numpy as np
from dateutil.relativedelta import relativedelta


time_series_path = sys.argv[1]
num_steps_ahead = int(sys.argv[2])
num_scenarios = int(sys.argv[3])


# charger le dataframe
df = pd.read_csv(time_series_path)

HORIZON = num_steps_ahead  # Predict the length of the test dataset

# convertir la colonne des dates en format de date
df['dates'] = pd.to_datetime(df['dates'])

# extraire l'année de chaque date
df['annee'] = df['dates'].dt.year

# trouver la valeur maximale de chaque année et calculer la proportion pour chaque mois
max_values = df.groupby(['annee']).max()

proportions = df.iloc[:, 1:5].div(max_values.iloc[:, 1:5].values.repeat(
    12, axis=0)[:df.shape[0]])  # diviser les valeurs par les maxima

# calculer la moyenne des proportions pour chaque mois
proportions.index = pd.to_datetime(df.iloc[:, 0], format='%Y-%m-%d')
means = proportions.groupby(proportions.index.month).mean()

# préparer les données pour la régression linéaire
X = max_values.index.values.reshape(-1, 1)
y = max_values.iloc[:, 1:5].values

# Créer des caractéristiques polynomiales de degré 2
poly = PolynomialFeatures(degree=2)
# Appliquer les caractéristiques polynomiales à X
X_poly = poly.fit_transform(X)

predicted_max_values = []

for year in range(int(np.ceil(num_steps_ahead / 12))):
    year_value = []
    for region_nb in range(4):
        y_region = y[:, region_nb]
        reg = LinearRegression().fit(X_poly, y_region)
        max_values_pred = max_values.index.max() + 1 + year
        X_pred = poly.transform(max_values_pred.reshape(-1, 1))
        max_value = reg.predict(X_pred)
        year_value.append(max_value[0])
    predicted_max_values.append(year_value)

df_predicted_max_value = pd.DataFrame(predicted_max_values).values.repeat(12, axis = 0)[:num_steps_ahead]

means_values = means.reset_index().iloc[:, 1:5]
means_values = pd.concat([means_values] * int(np.ceil(num_steps_ahead / 12)), ignore_index=True)

forecast_df = df_predicted_max_value * means_values

dates = [date + np.timedelta64(HORIZON, 'M') for date in df['dates'].iloc[-HORIZON:].values.astype('datetime64[M]')]

forecast_df.insert(0, 'dates',  dates)
forecast_df.insert(1, 'scenario', 1)

output_path = os.path.join(os.path.dirname(
    time_series_path), "_simulated_scenarios.csv")

# Write predictions to a CSV file
forecast_df.to_csv(output_path, index=False)
